<a href="https://www.kaggle.com/code/donkariuki/sasakazi-assesment?scriptVersionId=121017429" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# Import all required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import json

In [2]:
# Include file location for the dataset to be used
movie_dataset = "/kaggle/input/imdb-review-dataset" 

# List of all files in the dataset to be used 
movie_dataset_list = ["part-01.json", "part-02.json", "part-03.json", "part-04.json", "part-05.json", "part-06.json"]

# Initialize list to store the dataset on an empty list
imdb_movies = []

# Loop through the list of files and load the reviews into the list
for file in movie_dataset_list:
    with open(f"{movie_dataset}/{file}", mode='r') as movie_file:
        df_imdb_movies = json.load(movie_file)
        imdb_movies.extend(df_imdb_movies)

# Convert the list of movies to a pandas DataFrame
review_dataset = pd.DataFrame(imdb_movies)

In [3]:
review_dataset.head()

review_id         reviewer                          movie rating  \
0  rw5704482    raeldor-96879            After Life (2019– )      9   
1  rw5704483          dosleeb  The Valhalla Murders (2019– )      6   
2  rw5704484  brightconscious           Special OPS (2020– )      7   
3  rw5704485       gasconyway              #BlackAF (2020– )      8   
4  rw5704487     mmason-15867             The Droving (2020)      2   

          review_summary review_date  spoiler_tag  \
0   Very Strong Season 2  3 May 2020            0   
1  Icelandic detectives?  3 May 2020            0   
2        Nothing special  3 May 2020            0   
3               Good but  3 May 2020            0   
4       An honest review  3 May 2020            0   

                                       review_detail   helpful  
0  I enjoyed the first season, but I must say I t...    [1, 1]  
1  I know Iceland is a small country and police d...    [2, 2]  
2  Except K K , no other actor looks comfortable ...    [0, 0]  
3  I'm guessing that as a 62 year old white woman...    [5, 9]  
4  Here's the truth. There's not much to this mov...  [26, 41]

In [4]:
#inspecting data
review_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5571499 entries, 0 to 5571498
Data columns (total 9 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   review_id       object
 1   reviewer        object
 2   movie           object
 3   rating          object
 4   review_summary  object
 5   review_date     object
 6   spoiler_tag     int64 
 7   review_detail   object
 8   helpful         object
dtypes: int64(1), object(8)
memory usage: 382.6+ MB


In [5]:
review_dataset.tail()

review_id       reviewer               movie rating  \
5571494  rw0099137     Hairy_Lime    Sami swoi (1967)     10   
5571495  rw0099138   MuffinMan747    Sami swoi (1967)      3   
5571496  rw0099139  TYLERdurden74  Le Samouraï (1967)      9   
5571497  rw0099140         Mankin  Le Samouraï (1967)   None   
5571498  rw0099141      the reach  Le Samouraï (1967)     10   

                                            review_summary       review_date  \
5571494                         Funniest Polish Movie Ever  26 November 2003   
5571495                                           Horrible   8 February 2004   
5571496                                     Cult Movies 38   9 November 1998   
5571497      Some people seem to like this a lot, but why?      13 June 2000   
5571498  a unique import of the samurai code into the g...      16 July 2000   

         spoiler_tag                                      review_detail  \
5571494            0  Actually, the first three or four times I saw ...   
5571495            0  This movie glorifies the worst stereotypes of ...   
5571496            0  38. LE SAMOURAI (French-action, 1967) Jef (Ala...   
5571497            0  I found "Le Samourai" (**) to be more about st...   
5571498            0  J.P Mellville imports the samurai code of hono...   

          helpful  
5571494  [31, 34]  
5571495   [5, 82]  
5571496    [2, 8]  
5571497  [20, 41]  
5571498    [3, 9]

In [6]:
review_dataset.describe()

spoiler_tag
count  5.571499e+06
mean   2.129788e-01
std    4.094128e-01
min    0.000000e+00
25%    0.000000e+00
50%    0.000000e+00
75%    0.000000e+00
max    1.000000e+00

In [7]:
review_dataset.columns

Index(['review_id', 'reviewer', 'movie', 'rating', 'review_summary',
       'review_date', 'spoiler_tag', 'review_detail', 'helpful'],
      dtype='object')

In [8]:
#Dropping irellevant columns
review_dataset= review_dataset.drop(['review_date', 'spoiler_tag', 'helpful'], axis= 1)

In [9]:
review_dataset

review_id         reviewer                          movie rating  \
0        rw5704482    raeldor-96879            After Life (2019– )      9   
1        rw5704483          dosleeb  The Valhalla Murders (2019– )      6   
2        rw5704484  brightconscious           Special OPS (2020– )      7   
3        rw5704485       gasconyway              #BlackAF (2020– )      8   
4        rw5704487     mmason-15867             The Droving (2020)      2   
...            ...              ...                            ...    ...   
5571494  rw0099137       Hairy_Lime               Sami swoi (1967)     10   
5571495  rw0099138     MuffinMan747               Sami swoi (1967)      3   
5571496  rw0099139    TYLERdurden74             Le Samouraï (1967)      9   
5571497  rw0099140           Mankin             Le Samouraï (1967)   None   
5571498  rw0099141        the reach             Le Samouraï (1967)     10   

                                            review_summary  \
0                                     Very Strong Season 2   
1                                    Icelandic detectives?   
2                                          Nothing special   
3                                                 Good but   
4                                         An honest review   
...                                                    ...   
5571494                         Funniest Polish Movie Ever   
5571495                                           Horrible   
5571496                                     Cult Movies 38   
5571497      Some people seem to like this a lot, but why?   
5571498  a unique import of the samurai code into the g...   

                                             review_detail  
0        I enjoyed the first season, but I must say I t...  
1        I know Iceland is a small country and police d...  
2        Except K K , no other actor looks comfortable ...  
3        I'm guessing that as a 62 year old white woman...  
4        Here's the truth. There's not much to this mov...  
...                                                    ...  
5571494  Actually, the first three or four times I saw ...  
5571495  This movie glorifies the worst stereotypes of ...  
5571496  38. LE SAMOURAI (French-action, 1967) Jef (Ala...  
5571497  I found "Le Samourai" (**) to be more about st...  
5571498  J.P Mellville imports the samurai code of hono...  

[5571499 rows x 6 columns]

In [10]:
# Change the data type of rating  column from object to float 
review_dataset.rating = pd.to_numeric(review_dataset.rating)

In [11]:
#confirming the change from object to float
review_dataset.dtypes

review_id          object
reviewer           object
movie              object
rating            float64
review_summary     object
review_detail      object
dtype: object

## Question 1: Which movie has the highest rating?

In [12]:
# Sort the dataframe in descending order of rating column to show ratings from highest to lowest
review_dataset.sort_values(by='rating', ascending=False, inplace=True)
review_dataset.head()

review_id         reviewer                                movie  \
2639616  rw0892085       moondust17                  The Punisher (2004)   
2432038  rw2282417     ShadeGrenade  I Didn't Know You Cared (1975–1979)   
2432033  rw2282412    Crossfire3636                     Inception (2010)   
2432031  rw2282409  DonnieDarko6666                   Srpski film (2010)   
2432030  rw2282408       bosu_rares                         Adela (1985)   

         rating                                     review_summary  \
2639616    10.0                                        Great Movie   
2432038    10.0                                 Its grim up North!   
2432033    10.0  Can a movie get any better than being an exper...   
2432031    10.0  just a few thoughts on the movie "A Serbian Film"   
2432030    10.0                                A Beautiful Romance   

                                             review_detail  
2639616  This was a great movie.  I love movies when th...  
2432038  'I Didn't Know You Cared' crept unheralded int...  
2432033  Inception was one of those movies after which ...  
2432031  I've seen " A Serbian Film" twice now in betwe...  
2432030  I know it sounds simple, but the word fully de...

 There are more than one movie with the highest rating which is a 10.0. Next is to select the movie with the most number of the highest rating of 10.0

In [13]:
#  Selecting the movie with the most number of ratings by grouping them.
rating_number = review_dataset.groupby(['movie', 'rating']).size().reset_index(name='total_review_ratings')
rating_number.sort_values(by=['rating','total_review_ratings'],ascending=[False,False],inplace=True)

rating_number.head(1)

movie  rating  total_review_ratings
225159  Dil Bechara (2020)    10.0                  7188

##### As observed from above the movie with the highest total review ratings is Dil Bechara(2020) at rating of 10.0 with total number of ratings of 7188

## Question 2: How many reviews gave it the highest rating?

In [14]:
highest_rated_movie = rating_number.nlargest(1,'total_review_ratings')['movie'].values[0]
highest_rated_movie

'Dil Bechara (2020)'

In [15]:
highest_rating = rating_number.loc[rating_number['movie'] == highest_rated_movie, 'rating'].max()
highest_rating


10.0

In [16]:
# Count the number of unique reviewers who gave the top movie the highest rating
reviewers_count = review_dataset[(review_dataset['movie'] == highest_rated_movie) & (review_dataset['rating'] == highest_rating)].reviewer.nunique()
print(f"{reviewers_count} unique reviewers gave the highest rating")

7132 unique reviewers gave the highest rating


## Question 3: what is the least rating and review details?

In [17]:
#finding the least rating ofthe highest rated movie
least_rating = review_dataset.loc[review_dataset['movie'] == highest_rated_movie, 'rating'].min()
least_rating

1.0

In [18]:
least_rated_movie = review_dataset[(review_dataset['movie'] == highest_rated_movie) & (review_dataset['rating'] == least_rating)]
least_rated_movie.tail(1)

review_id    reviewer               movie  rating  \
610690  rw5955538  dush-k1996  Dil Bechara (2020)     1.0   

                               review_summary  \
610690  Love for Sushant but a terrible movie   

                                            review_detail  
610690  The movie seems kinda rushed and scenes seem r...

In [19]:
rating_least_rated_movie = least_rated_movie['rating'].values[0]
review_detail = least_rated_movie['review_detail'].values[0]
print(f"Rating of the movie {highest_rated_movie} is: {rating_least_rated_movie}")
print(f"Review details of {highest_rated_movie} is : {review_detail}")

Rating of the movie Dil Bechara (2020) is: 1.0
Review details of Dil Bechara (2020) is : Well movie isn't that good as people are saying. They are just showing sympathy that's all. surely not worth a watch. Music direction script all failures. I guess people should treat it as a movie and of they wanna give ratings then give them to those who deserves.
